### PR TITLE
Backoffice actions behavior update

### LINKED ISSUE #1317: [BAL-301] Backoffice actions behavior updates
### Description

* Update a single document from the backoffice
* Document validation on approval

<sub>From [SyncLinear.com](https://synclinear.com) | [BAL-301](https://linear.app/ballerine/issue/BAL-301/backoffice-actions-behavior-updates)</sub>

### PR DESCRIPTION
**Pull request by alonp99**
### Description
* Update a single document from the backoffice
* Document validation on approval

### CODE REVIEW

File: services/workflows-service/src/events/document-changed-webhook-caller.ts [hunk 1]
```diff
@@ -24,9 +24,11 @@ export class DocumentChangedWebhookCaller {
   ) {
     this.#__axios = this.httpService.axiosRef;
 
-    workflowEventEmitter.on('workflow.context.changed', async data => {
+    // @ts-ignore
+    workflowEventEmitter.on('workflow.context.changed', () => {
       try {
-        await this.handleWorkflowEvent(data);
+        console.log('skipping webhook');
```
Comment (alonp99) [inline services/workflows-service/src/events/document-changed-webhook-caller.ts]: remove before merge

**Review by ballerine-dev — APPROVED** [review]

File: apps/backoffice-v2/src/domains/entities/hooks/mutations/useApproveTaskByIdMutation/useApproveTaskByIdMutation.tsx [hunk 2]
```diff
@@ -59,7 +59,8 @@ export const useApproveTaskByIdMutation = (workflowId: string) => {
       toast.success(t('toast:approve_document.success'));
     },
     onError: (_error, _variables, context) => {
-      toast.error(t('toast:approve_document.error'));
+      console.log({ _error, _variables, context });
```
Comment (MatanYadaev) [inline apps/backoffice-v2/src/domains/entities/hooks/mutations/useApproveTaskByIdMutation/useApproveTaskByIdMutation.tsx:62]: Leftover - should be removed

File: apps/backoffice-v2/src/pages/Entity/components/EditableDetails/EditableDetails.tsx [hunk 3]
```diff
@@ -59,46 +59,43 @@ export const EditableDetails: FunctionComponent<IEditableDetails> = ({
...
     defaultValues,
   });
-  const { mutate: mutateUpdateWorkflowById } = useUpdateWorkflowByIdMutation({
+  const { mutate: mutateUpdateWorkflowById } = useUpdateDocumentByIdMutation({
     workflowId,
+    documentId: valueId,
   });
   const onMutateTaskDecisionById = ({
-    context,
+    // Ask Omri about this
```
Comment (MatanYadaev) [inline apps/backoffice-v2/src/pages/Entity/components/EditableDetails/EditableDetails.tsx:67]: ?

File: apps/backoffice-v2/src/pages/Entity/components/EditableDetails/interfaces.ts [hunk 4]
```diff
@@ -16,6 +16,7 @@ export interface IEditableDetails {
   valueId: string;
   id: string;
   documents: Array<{
+    propertiesSchema: any;
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/pages/Entity/components/EditableDetails/interfaces.ts:19]: This type should be safe to use here.
```suggestion
propertiesSchema: Record<string, unknown>;
```

File: services/workflows-service/src/workflow/dtos/document-update-update-input.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import { ApiProperty } from '@nestjs/swagger';
+import { IsObject } from 'class-validator';
+
+export class DocumentUpdateInput {
+  @ApiProperty({
+    required: true,
+    type: Object,
+  })
+  @IsObject()
+  document!: any;
```
Comment (Omri-Levy) [inline services/workflows-service/src/workflow/dtos/document-update-update-input.ts:10]: Until Typebox is in for documents.
```suggestion
document!: Record<string, unknown>;
```

File: services/workflows-service/src/workflow/workflow.service.ts [hunk 6]
```diff
@@ -648,38 +611,124 @@ export class WorkflowService {
...
-        };
+    const documentWithDecision = {
+      ...document,
+      id: document.id,
+      decision: {
+        ...newDecision,
+        status,
       },
+    };
+    const checkRequiredFields = status === 'approved' ? true : false;
```
Comment (Omri-Levy) [inline services/workflows-service/src/workflow/workflow.service.ts:626]: Can be simplified.
```suggestion
const checkRequiredFields = status === 'approved';
```

File: services/workflows-service/src/workflow/workflow.service.ts [hunk 7]
```diff
@@ -648,38 +611,124 @@ export class WorkflowService {
...
+    {
+      workflowId,
+      documentId,
+      checkRequiredFields = true,
+    }: {
+      workflowId: string;
+      documentId: string;
+      checkRequiredFields?: boolean;
+    },
+    data: DefaultContextSchema['documents'][number] & { propertiesSchema?: object },
```
Comment (Omri-Levy) [inline services/workflows-service/src/workflow/workflow.service.ts:644]: 90% of the time using `Record` over `object` is the way to go. `object` disallows using properties on the object (unless runtime checks are made).

File: services/workflows-service/src/workflow/workflow.service.ts [hunk 8]
```diff
@@ -648,38 +611,124 @@ export class WorkflowService {
...
+        updatedWorkflow.context?.documents?.every(
+          (document: DefaultContextSchema['documents'][number]) => {
+            return ['approved', 'rejected', 'revision'].includes(
+              document?.decision?.status as string,
+            );
+          },
+        );
+
+      updatedWorkflow.status = allDocumentsResolved ? 'completed' : updatedWorkflow.status;
+      await this.workflowRuntimeDataRepository.updateById(workflowId, {
```
Comment (Omri-Levy) [inline services/workflows-service/src/workflow/workflow.service.ts:698]: Using the return value of `updateById` over mutating `updatedWorkflow` seems safer for the long term, over time `updateById` may change data differently.

File: services/workflows-service/src/workflow/workflow.service.ts [hunk 9]
```diff
@@ -648,38 +611,124 @@ export class WorkflowService {
...
+      id: documentId,
+    };
+
+    const newDocument = addPropertiesSchemaToDocument(document);
+    const propertiesSchema = newDocument?.propertiesSchema ?? {};
+    if (Object.keys(propertiesSchema)?.length) {
+      let propertiesSchemaForValidation = propertiesSchema;
+      if (!checkRequiredFields) {
+        const { required: _required, ..._propertiesSchemaForValidation } = propertiesSchema;
+        propertiesSchemaForValidation = _propertiesSchemaForValidation;
```
Comment (Omri-Levy) [inline services/workflows-service/src/workflow/workflow.service.ts:661]: Using `_` makes it seem like the value is not in use.
```suggestion
propertiesSchemaForValidation = restPropertiesSchemaForValidation
```

File: services/workflows-service/src/workflow/workflow.service.ts [hunk 10]
```diff
@@ -648,38 +611,124 @@ export class WorkflowService {
...
-
     const correlationId = await this.getCorrelationIdFromWorkflow(updatedWorkflow);
 
+    if (
+      ['active'].includes(updatedWorkflow.status) &&
+      workflowDef.config?.completedWhenTasksResolved
+    ) {
+      const allDocumentsResolved =
+        updatedWorkflow.context?.documents?.length &&
+        updatedWorkflow.context?.documents?.every(
```
Comment (Omri-Levy) [inline services/workflows-service/src/workflow/workflow.service.ts:689]: Note we have an helper called `safeEvery` to cover the `length` + `every` pairing.

File: apps/backoffice-v2/src/domains/workflows/hooks/mutations/useUpdateDocumentByIdMutation/useUpdateDocumentByIdMutation.tsx [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+      queryClient.setQueryData(workflowById.queryKey, (oldWorkflow: TWorkflowById) => {
+        return {
+          ...oldWorkflow,
+          context: {
+            ...oldWorkflow.context,
+            documents: oldWorkflow.context.documents.map((doc: { id: string }) => {
+              if (doc.id === documentId) {
+                return document;
+              }
+              return doc;
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/domains/workflows/hooks/mutations/useUpdateDocumentByIdMutation/useUpdateDocumentByIdMutation.tsx:48]: Spreading here probably fits optimistic updates better.

File: apps/backoffice-v2/src/pages/Entity/components/EditableDetails/EditableDetails.tsx (hunk 3 above)
Comment (Omri-Levy) [inline apps/backoffice-v2/src/pages/Entity/components/EditableDetails/EditableDetails.tsx:67]: 👀

File: apps/backoffice-v2/src/pages/Entity/components/EditableDetails/EditableDetails.tsx [hunk 12]
```diff
@@ -59,46 +59,43 @@ export const EditableDetails: FunctionComponent<IEditableDetails> = ({
...
     mutateUpdateWorkflowById({
-      context,
+      document,
       action,
     });
   const onSubmit: SubmitHandler<Record<PropertyKey, unknown>> = formData => {
-    const context = {
-      documents: documents?.map(document => {
-        if (document?.id !== valueId) return document;
+    const document = documents?.find(document => document?.id === valueId);
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/pages/Entity/components/EditableDetails/EditableDetails.tsx:79]: Do we need to use the entire document here over just type, category, and properties?

**Pull request merged**